grpc-proto/src/main/java/io/stargate/grpc/VIntCoding.java

```diff
@@ -0,0 +1,181 @@
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+package io.stargate.grpc;
+
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.IOException;
+
+/**
+ * Variable length encoding inspired from Google <a
+ * href='https://developers.google.com/protocol-buffers/docs/encoding#varints'>varints</a>.
+ *
+ * <p>Cassandra vints are encoded with the most significant group first. The most significant byte
+ * will contains the information about how many extra bytes need to be read as well as the most
+ * significant bits of the integer. The number of extra bytes to read is encoded as 1 bit on the
+ * left side. For example, if we need to read 3 more bytes the first byte will start with 1110. If
+ * the encoded integer is 8 bytes long the vint will be encoded on 9 bytes and the first byte will
+ * be: 11111111
+ *
+ * <p>Signed integers are (like protocol buffer varints) encoded using the ZigZag encoding so that
+ * numbers with a small absolute value have a small vint encoded value too.
+ *
+ * <p>Note that there is also a type called {@code varint} in the CQL protocol specification. This
+ * is completely unrelated.
+ */
+class VIntCoding {
+
+  private static long readUnsignedVInt(DataInput input) throws IOException {
+    int firstByte = input.readByte();
+
+    // Bail out early if this is one byte, necessary or it fails later
+    if (firstByte >= 0) {
+      return firstByte;
+    }
+
+    int size = numberOfExtraBytesToRead(firstByte);
+    long retval = firstByte & firstByteValueMask(size);
+    for (int ii = 0; ii < size; ii++) {
+      byte b = input.readByte();
+      retval <<= 8;
+      retval |= b & 0xff;
+    }
+
+    return retval;
+  }
+
+  static long readVInt(DataInput input) throws IOException {
+    return decodeZigZag64(readUnsignedVInt(input));
+  }
+
+  // & this with the first byte to give the value part for a given extraBytesToRead encoded in the
+  // byte
+  private static int firstByteValueMask(int extraBytesToRead) {
+    // by including the known 0bit in the mask, we can use this for encodeExtraBytesToRead
+    return 0xff >> extraBytesToRead;
+  }
+
+  private static byte encodeExtraBytesToRead(int extraBytesToRead) {
+    // because we have an extra bit in the value mask, we just need to invert it
+    return (byte) ~firstByteValueMask(extraBytesToRead);
+  }
+
+  private static int numberOfExtraBytesToRead(int firstByte) {
+    // we count number of set upper bits; so if we simply invert all of the bits, we're golden
+    // this is aided by the fact that we only work with negative numbers, so when upcast to an int
+    // all
+    // of the new upper bits are also set, so by inverting we set all of them to zero
+    return Integer.numberOfLeadingZeros(~firstByte) - 24;
+  }
+
+  private static final ThreadLocal<byte[]> encodingBuffer =
```

Comment:
Is it worth having this for a tiny 9 byte buffer? Yes, it reduces churn a bit but not sure it's a net win due to need for `ThreadLocal` access.

Reply:
If you do `final byte[] enocodingBuffer = new byte[9]` locally and pass it into `encodeVInt()` maybe it'll inline that method then escape analysis will make that a stack alloc? I don't know how to go about analyzing that. :)

Reply:
Even without that I think that allocation (and GC) of 9-byte array is trivially small thing, not worth worrying about unless in a tight loop or something.
In fact this is kind of thing that I think really should not be optimized unless it shows up in profiling, heap dump or something.

Reply:
Yeah that seems a bit overkill, I've removed it.